core/src/main/java/org/apache/iceberg/encryption/StandardEncryptionManager.java

```diff
@@ -0,0 +1,120 @@
+  private String tableKeyId;
+  private int dataKeyLength;
+  private boolean kmsGeneratedKeys;
+
+  private transient volatile SecureRandom workerRNG = null;
+
+  /**
+   * @param tableKeyId table encryption key id
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   * @param encryptionProperties encryption properties
+   */
+  public StandardEncryptionManager(
+      String tableKeyId, KeyManagementClient kmsClient, Map<String, String> encryptionProperties) {
+    Preconditions.checkNotNull(
+        tableKeyId,
+        "Cannot create EnvelopeEncryptionManager because table encryption key ID is not specified");
+    Preconditions.checkNotNull(
+        kmsClient, "Cannot create EnvelopeEncryptionManager because kmsClient is null");
+    Preconditions.checkNotNull(
+        encryptionProperties,
+        "Cannot create EnvelopeEncryptionManager because encryptionProperties are not passed");
+    this.tableKeyId = tableKeyId;
+    this.kmsClient = kmsClient;
+    this.kmsGeneratedKeys = kmsClient.supportsKeyGeneration();
+
+    this.dataKeyLength =
+        PropertyUtil.propertyAsInt(
+            encryptionProperties,
+            EncryptionProperties.ENCRYPTION_DEK_LENGTH,
+            EncryptionProperties.ENCRYPTION_DEK_LENGTH_DEFAULT);
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    if (null == workerRNG) {
+      createSecureRandomGenerator();
+    }
+
+    ByteBuffer fileDek = ByteBuffer.allocate(dataKeyLength);
+    workerRNG.nextBytes(fileDek.array());
+
+    ByteBuffer aadPrefix = ByteBuffer.allocate(EncryptionProperties.ENCRYPTION_AAD_LENGTH_DEFAULT);
+    workerRNG.nextBytes(aadPrefix.array());
+
+    // For data files
+    KeyMetadata dataEncryptionMetadata = new KeyMetadata(fileDek, null, aadPrefix);
+
+    // For metadata files
+    // This is an expensive operation, RPC to KMS server
+    ByteBuffer wrappedDek = kmsClient.wrapKey(fileDek, tableKeyId);
+    KeyMetadata manifestEncryptionMetadata = new KeyMetadata(wrappedDek, tableKeyId, aadPrefix);
+
+    // We don't know which key metadata to use, because we don't know what file we encrypt.
+    // Works for data files:
+    // return new BaseEncryptedOutputFile(new AesGcmOutputFile(rawOutput, fileDek, aadPrefix),
+    //    dataEncryptionMetadata, rawOutput);
+    // Works for manifest files:
+    // return new BaseEncryptedOutputFile(new AesGcmOutputFile(rawOutput, fileDek, aadPrefix),
+    //    manifestEncryptionMetadata, rawOutput);
+
+    // Temp return, for parquet data only. TODO - remove
+    return new BaseEncryptedOutputFile(null, dataEncryptionMetadata, rawOutput);
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    if (encrypted.keyMetadata() == null || encrypted.keyMetadata().buffer() == null) {
+      throw new RuntimeException(
+          "Unencrypted file " + encrypted.encryptedInputFile().location() + " in encrypted table");
```

Comment:
Shouldn't use `RuntimeException` in general. Is there a better one?

Also, is this an exception? What happens when converting a table to use encryption? It seems like this could be a valid state.

Reply:
Sometimes, due to misconfiguration etc - users think they write encrypted tables / files, but the result is not encrypted. So this is a rather important security check. But we can discuss the options and trade-offs between this and a flexibility to be able to convert an existing table to use encryption.

Reply:
The check I've mentioned, is not always possible to perform. Switching to your suggestion.